### DISCUSSION TITLE
Shortcut improvement suggestions

### DISCUSSION THREAD

**Discussion by mikkovedru** [Ulauncher/Ulauncher#1377] [Ideas]
### Environment

Linux Mint 21.3 Cinnamon (6.0.4)

### Bug description

These issues are present in both v5 and the latest 6.0.0-beta6.

## Problem 1
I created shortcuts, but queries did not complete as expected. The keywords I typed were not there. Trying to search DuckDuckGo would throw me to the literal page of `https://duckduckgo.com/?q=%s` (notice the `%s`).
![image](https://github.com/Ulauncher/Ulauncher/assets/2521942/98ffa626-a836-4fd4-ac96-8f70332e0574)

Turns out that this happens because of the enabled option "Run without arguments (Allows you to type in a keyword and press Enter to run a shortcut)".
![image](https://github.com/Ulauncher/Ulauncher/assets/2521942/7367c07c-65f4-4f49-b2b8-13aceb680072)

This is really not obvious to me (a very experienced software user and developer)! This is how I understood the option:
- If enabled, you can type merely the main keyword and it will run/open the website/script. If you type other keywords, those keywords will also be forwarded.
- If disabled, merely typing the main keyword would do nothing (maybe close the window or demand more keywords be typed), until you type some more. So running empty scripts without keywords would not work; keywords are required.

The current behavior is not obvious and **feels** like a bug. It makes very little sense to have a separate option for such a border case of "if the query/script absolutely must have `%s` in it".

Whether it is indeed "a bug" that needs to be fixed or whether the solution is in writing a better explanation - it's for you to decide.

## Problem 2
When writing the keywords, the keyword and space are being shows.
![image](https://github.com/Ulauncher/Ulauncher/assets/2521942/67463375-ff8c-4d19-b9c7-bc75bee93d07)
This gives the wrong impression of the query that will be run after pressing Enter.

This is the problem regardless of the status of the "Run without arguments" option.

A keyword and space should be sanitized away **if appropriate**. The preview should show the command that actually is going to be executed.

## Problem 3
"Default search (Suggest this shortcut when no results found)" also doesn't work the way I - as a user - would imagine it should.

1. If these queries should be available when there are no results, why is this list of queries not available immediately after running Ulauncher and it being empty? It should.
2. When starting to write, this list disappears. Why? This is bad because:
	2.1. I would like to be able to invoke queries at any point of the process. "Ok, I am trying to find something, oh, never mind, I will just google it".
	2.2. The moment when this list appears again is really unpredictable. This is because of the search function and it is trying to find stuff that is only closely related to the inquiry, instead of searching for only identical strings. Look at this craziness:
![image](https://github.com/Ulauncher/Ulauncher/assets/2521942/3ca1e94a-9261-426e-99b5-6cf05111880e)
I am not saying that this is a bad thing by itself. No, it's good that the fuzzy search is trying to be helpful and find something at all. But the issue is that this behavior makes it totally unpredictable when the Ulauncher finally decides that enough is enough and will give me the list of default queries. It can take many more key presses after the I understood that I don't have the file and must run the query instead. Unpredictability is a big problem!
3. Hotkeys (Alt+1, etc). 
![image](https://github.com/Ulauncher/Ulauncher/assets/2521942/b1ca4fb1-fe5a-41cb-ace1-f69540fe6b92)
But what happens when I start typing?
![image](https://github.com/Ulauncher/Ulauncher/assets/2521942/bedccfa0-5a09-47ea-931b-c7c173b9e3ef)
Oops! Now the hotkeys for my default queries are totally changed. Not good! Now I cannot memorize the main hotkeys and can't rely on them.

## Suggestions/fixes
1. At the very least, should improve text of the Problem 1.
2. Perhaps get "the bug" fixed by changing the functionality from the current one to the one I suggested. If not, the current implementation can IMHO be deleted altogether. I might be wrong, but I personally can't think of any good reason to have this current implementation present at all; a rare border case is not worthy to be cluttering the UI and making the program unnecessarily more complex.
3. The preview should show the command that is actually going to be executed. A keyword and space should be sanitized away **if appropriate**.
4. Related to the list of default search:
	4.1. Make the list **always** visible, from the first moment I run the program. Perhaps a toggle to disable this feature, even though I personally think that it is not necessary. When starting to write arguments, those should be a priority; so display the default list on the bottom (show the whole list starting from the bottom of the screen, and then fill the rest of the empty space above with the search results).
	4.2. Make the currently selected query visually recognizable from other default queries. And those, in turn, should be visually different from normal text/arguments.
	4.3. Even if one writes keyword hotkey (`g`), other default searches should still be visible. Only visually accentuate the one default query that is being currently chosen.
	4.4. When typing the keyword and starting to write arguments, still show all 3 options (from top of the screen to the bottom): (1) the main default search chosen via the specific hotkey keyword; (2) ignoring the hotkey, what would the local search show?; (3) All other default queries. All 3 types should be visually different from one another.
	4.5. Reserve the hotkeys for each default search queries. When more keywords are being written, give away only freely available Alt-hotkeys.
	4.6. When starting to write something (like `g test`), allow choosing another default query. In this case, delete the current keyword and add the one of the new selected query.

- Bonus 1: Make the list of queries manually sortable. So by changing places/importance, one could change the hotkey number for whatever query (`Alt + 1`, etc).
- Bonus 2: have 2 separate queries/scripts: for commands with arguments and for commands without. Thus, one could e.g. both search on Wikipedia, but also quickly go to my personal wiki page (or list of monitored pages).

Tell me if there is confusion, so I can try to explain even better.

### Communication guidelines

- [X] I have read and agree with Ulauncher's communication guidelines and checked the troubleshooting page for self-help.
- [X] I have tried to understand the error message and help myself, before reporting it.
- [X] My issue description has clear steps for how to reproduce the issue.
- [X] This issue applies to the whole Ulauncher application, not for a specific extension.

### Ulauncher v6

- [X] I have made sure the bug isn't already fixed in v6.

**Comment by friday** [comment]
It would take me longer to respond to this issue than it would to make the changes. I agree in part and will look at improving some of these issues, but it's low prio and some of the things you suggest is not even technically possible, and others would be breaking the current behavior.

Sorting shortcuts already has a task: https://github.com/Ulauncher/Ulauncher/issues/1008

If you want to be able to type any text and have Ulauncher parse that and show a list of compatible extensions and shortcuts, that's just not how it works. The "default search" suggestions are just an exception to show when there are no other results. I think that would be cool myself and I brought up that reverse flow for extensions here: https://github.com/Ulauncher/Ulauncher/discussions/966, but I don't think we'll ever get that far. It's too foreign with how Ulauncher is behaving currently. Just use the keyword instead.

If you want a global keyboard shortcut to trigger a ulauncher shortcut you should do that in your Desktop environment, but instead add: `gapplication action io.ulauncher.Ulauncher set-query "'d '"` (for v6). The jump keys were fundamentally not designed to be persistent per result. That would obviously not work since there is an infinite amount of results:
![image](https://github.com/Ulauncher/Ulauncher/assets/515120/11579d9e-79ad-454f-bfee-bc948b5ffa31)

**Comment by friday** [comment]
And also, removing `%s` would not solve the problem, because we also support shell script argument for scripts, and each language implements that differently. For example bash uses `$@` or `$1`. Nodejs uses `process.argv`. Python uses `sys.argv`. We can't support to detect this for all scripting languages.

The real problem imo is that shortcuts without arguments shouldn't have keywords either. Keywords are designed to trigger a "prompt" to get a user input, but we don't want that. They should work just like apps instead imo.

**Comment by friday** [comment]
#1376

**Comment by friday** [comment]
I'm going to convert this to a discussion now. Issues should be clear and actionable and not bring up multiple things in one issue.

**Reply by friday** [reply]
This should now be more clear: 5742f1e11bcfa93bb4e88ac24175016c660210d5 cc5b90b
